### DISCUSSION TITLE
[proposal] Howto: Vote In Codeowner Changes

### DISCUSSION THREAD

**Discussion by hcgatewood** [magma/magma#5524] [Ideas]
# Howto: Vote In Codeowner Changes

The process for updating codeowners is underspecified at the moment. This proposal puts forward a de facto method for updating codeowners while preserving the low-friction process.

### Conceptual requirements

Conceptually, updating codeowners (i.e. changing a line in the [CODEOWNERS.md](https://github.com/magma/magma/blob/master/CODEOWNERS) file), would require the following

- Majority of all existing codeowners vote yea, *and*
- Strictly less than 3 existing codeowners vote nay

Note: "all existing codeowners" is equal to the set of codeowners (preexisting or vote-approved) at the time the vote is initialized.

### Particular mechanism

Desired mechanism provides the following affordances

- *Ergonomic.* Easy to use, easy to verify, user-friendly, low-friction
- *Anonymous.* Vote anonymously, report only aggregate output
- *Time-bounded.* Automatically close the vote after set period of time
- *Immutable.* Votes locked in at deadline

## Proposal: Accord Slack app

[Accord](https://accord.coop/docs) is a Slack app that provides the above-mentioned affordances via the `/decide` Slack command.

The entire workflow can remain in the `#governance-codeowners` channel, along with satisfying the above desiderata.

### Workflow

See afterword for screenshots.

- A codeowner posts an informal codeowners update proposal in `#governance-codeowners`
- A TSC member ("nominator") adds a new poll via the `/decide` Slack command
    - Poll content
        - Title: informative summary
        - Description: specifics of the nomination, where exactly 1 username is under consideration for that nomination, and can include both additions and removals in the same nomination
            -  For additions: *1 username* that will be mapped to *1 or more directories*
            - For removals: *1 username* that will be removed from *1 or more directories*
        - Options: `Yes`, `No`, `Abstain`
    - Poll settings
        - Method: `approval` (require absolute number of yea votes)
        - Quorum: `anymone` (none)
        - Close voting in: `2 days` or `1 week`
        - Post in: `governance-codeowners`
        - Number of winners: `1`
        - Minimum approval: absolute number that constitutes a strict majority of current codeowners
- After the poll auto-closes, the nominator announces the outcome
    - Yea: `yes` votes totaling a strict majority *and* `no` votes strictly less than 3
        - Note: poll will say "`yes` won" even if there are more than 3 `no` votes, so the nominator must properly reconcile these values
    - Nay: default
- Nominator handles updating repo to match outcome
    - Open PR with changes to `CODEOWNERS.md`
    - Update `#governance-codeowners` channel membership (if necessary)
    - Update GitHub permissions (if necessary)
- Can re-initiate a particular poll if desired, but by convention wait a few weeks

### Evaluation

Pros

- Ergonomic: everything occurs directly in Slack
- Anonymous: individual votes not exposed
- Time-bounded: auto-closes and reports results, with minor manual effort needed to interpret outcome
- Immutable: can't change vote after poll closes
- Other: hard-coded rules about needing an absolute number of votes

Cons

- Can vote for multiple of yes/no/abstain. Probably not a big deal in practice
- Must manually reconcile the case where a majority of `yes` votes were received *and* 3 or more `no` votes received (i.e. interpreting this as a nay outcome)

### Voting in a poll

![Screen Shot 2021-03-15 at 9 30 40 PM](https://user-images.githubusercontent.com/8029544/111258349-cfcb9800-85ea-11eb-8442-7aef8ef18b88.png)
![Screen Shot 2021-03-15 at 9 31 01 PM](https://user-images.githubusercontent.com/8029544/111258355-d22df200-85ea-11eb-9906-2fd6098e5a56.png)
![Screen Shot 2021-03-15 at 10 08 17 PM](https://user-images.githubusercontent.com/8029544/111258835-c131b080-85eb-11eb-9ad3-78cdad1f8e07.png)

### Creating a poll

![Screen Shot 2021-03-15 at 9 21 36 PM](https://user-images.githubusercontent.com/8029544/111258283-b0346f80-85ea-11eb-9db2-e18d04566859.png)
![Screen Shot 2021-03-15 at 9 22 22 PM](https://user-images.githubusercontent.com/8029544/111258307-ba566e00-85ea-11eb-85b9-cbc8f3d07fe0.png)
![Screen Shot 2021-03-15 at 9 22 29 PM](https://user-images.githubusercontent.com/8029544/111258318-c04c4f00-85ea-11eb-833f-1f130303cc06.png)

**Comment by electronjoe** [comment]
In the `Workflow` example, is the process and wording clearly applicable for both adding and removing codeowners? I think it is - but it's not explicit. Would the poll description still contain the following.

> including 1 username that will be mapped to 1 or more directories

**Reply by hcgatewood** [reply]
Good call, I'll update

**Comment by themarwhal** [comment]
This looks good to me!
